Handbook: drastically improve I18N docs (#12576)
WordPress/gutenberg: docs/designers-developers/developers/internationalization.md

# Internationalization

This document aims to give an overview of the possibilities for both internationalization and localization when developing with Gutenberg.

## PHP

For years, WordPress has been providing the necessary tools and functions to internationalize plugins and themes. This includes helper functions like `__()` and similar.

### Common Methods

- `__( 'Hello World', 'my-text-domain' )`: Translate a certain string.
- `_x( 'Block', 'noun', 'my-text-domain' )`: Translate a certain string with some additional context.
- `_e( 'Hello World', 'my-text-domain' )`: Translate and print a certain string.

> - `esc_html__( 'Hello World', 'my-text-domain' )`: Translate a certain string and escape it for safe use in HTML output

Reviewer: ```suggestion
- `esc_html__( 'Hello World', 'my-text-domain' )`: Translate a certain string and escape it for safe use in HTML output.
```

> - `esc_html_e( $string_to_translate, 'my-text-domain' )`: Translate a certain string, escape it for safe use in HTML output, and print it.

Reviewer: ```suggestion
- `esc_html_e( 'Hello World', 'my-text-domain' )`: Translate a certain string, escape it for safe use in HTML output, and print it.
```

- `_n( '%s Comment', '%s Comments', $number, 'my-text-domain' )`: Translate and retrieve the singular or plural form based on the supplied number.  
  Usually used in combination with `sprintf()` and `number_format_i18n()`.

## JavaScript

Historically, `wp_localize_script()` has been used to put server-side PHP data into a properly-escaped native JavaScript object.

The new editor introduces a new approach to translating strings for the editor through a new package called `@wordpress/i18n`.

The new script package is registered with WordPress as `wp-i18n` and should be declared as a dependency during `wp_register_script()` and imported as a global off the Window object as `wp.i18n`.

> Depending on your developer workflow, you might want to use `wp i18n make-pot` or a build tool for Babel called `@wordpress/babel-plugin-makepot` to create the necessary translation file. The latter which requires use of Babel to compile code to extract the i18n methods.

Reviewer: It's not clear that `wp i18n make-pot` is a WP-CLI command. Maybe add something like this:

```suggestion
Depending on your developer workflow, you might want to use WP-CLI's `wp i18n make-pot` command or a build tool for Babel called `@wordpress/babel-plugin-makepot` to create the necessary translation file. The latter which requires use of Babel to compile code to extract the i18n methods.
```
